### PR TITLE
SONAR-8332 add http request ID

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* CRITICAL 3 critical

Watch the comments in this conversation to review them.

File: sonar-core/src/main/java/org/sonar/core/util/UuidGenerator.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * See https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java
+ * </p>
+ */
+public interface UuidGenerator {
+  byte[] generate();
+
+  WithFixedBase withFixedBase();
+
+  @FunctionalInterface
+  interface WithFixedBase {
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/util/UuidGenerator.java]: MAJOR Drop this interface in favor of "java.util.function.IntFunction<$Array>". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1711)

File: sonar-core/src/test/java/org/sonar/core/util/UuidGeneratorImplTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+public class UuidGeneratorImplTest {
+  private UuidGeneratorImpl underTest = new UuidGeneratorImpl();
+
+  @Test
+  public void generate_returns_unique_values_without_common_initial_letter_given_more_than_one_milisecond_between_generate_calls() throws InterruptedException {
+    Base64.Encoder encoder = Base64.getEncoder();
+    int count = 30;
+    Set<String> uuids = new HashSet<>(count);
+    for (int i = 0; i < count; i++) {
+      Thread.sleep(5);
```
Comment (SonarTech) [inline sonar-core/src/test/java/org/sonar/core/util/UuidGeneratorImplTest.java]: MAJOR Remove this use of "Thread.sleep()". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2925)

File: sonar-core/src/main/java/org/sonar/core/util/UuidGeneratorImpl.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    @Override
+    public byte[] get() {
+      int sequenceId = getSequenceId();
+      initBase(sequenceId);
+      return super.generate(sequenceId);
+    }
+  }
+
+  private class FixedBasedUuidGenerator extends UuidGeneratorBase implements WithFixedBase {
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/util/UuidGeneratorImpl.java]: MAJOR Make this a "static" inner class. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2694)

File: sonar-core/src/main/java/org/sonar/core/util/UuidGeneratorImpl.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    /** Puts the lower numberOfLongBytes from l into the array, starting index pos. */
+    private static void putLong(byte[] array, long l, int pos, int numberOfLongBytes) {
+      for (int i = 0; i < numberOfLongBytes; ++i) {
+        array[pos + numberOfLongBytes - i - 1] = (byte) (l >>> (i * 8));
+      }
+    }
+  }
+
+  private final class FullNewUuidGenerator extends UuidGeneratorBase implements Supplier<byte[]> {
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/util/UuidGeneratorImpl.java]: MAJOR Make this a "static" inner class. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2694)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* CRITICAL 2 critical
* MAJOR 4 major

Watch the comments in this conversation to review them.

File: sonar-core/src/main/java/org/sonar/core/util/UuidGenerator.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  byte[] generate();
+
+  /**
+   * Provide a new {@link UuidGenerator} instance which will allow generation of UUIDs which base is constant and can
+   * vary according to a provided increment value (see {@link WithFixedBase#generate(int)}).
+   */
+  WithFixedBase withFixedBase();
+
+  @FunctionalInterface
+  interface WithFixedBase {
```
Comment (SonarTech) [inline sonar-core/src/main/java/org/sonar/core/util/UuidGenerator.java]: MAJOR Drop this interface in favor of "java.util.function.IntFunction<$Array>". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1711)

File: server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestUidGeneratorBase.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.platform.web.requestid;
+
+import org.sonar.core.util.UuidGenerator;
+
+public interface RequestUidGeneratorBase {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestUidGeneratorBase.java]: CRITICAL Annotate the "RequestUidGeneratorBase" interface with the @FunctionalInterface annotation [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1609)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* CRITICAL 3 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: sonar-core/src/main/java/org/sonar/core/util/UuidGenerator.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+ * </p>
+ */
+public interface UuidGenerator {
+  /**
+   * Generates a UUID which base and increment are always different from any other value provided by this method.
+   */
+  byte[] generate();
+
+  /**
+   * Provide a new {@link UuidGenerator} instance which will allow generation of UUIDs which base is constant and can
```
Comment (simonbrandhof) [inline sonar-core/src/main/java/org/sonar/core/util/UuidGenerator.java]: To be exact `{@link UuidGenerator}` should be replaced by `UUID generator`.

File: sonar-core/src/test/java/org/sonar/core/util/UuidGeneratorImplTest.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.Set;
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class UuidGeneratorImplTest {
+  private UuidGeneratorImpl underTest = new UuidGeneratorImpl();
+
+  @Test
+  public void generate_returns_unique_values_without_common_initial_letter_given_more_than_one_milisecond_between_generate_calls() throws InterruptedException {
```
Comment (simonbrandhof) [inline sonar-core/src/test/java/org/sonar/core/util/UuidGeneratorImplTest.java]: On the opposite the first characters of sequentially generated UUIDs share a common prefix. This is an optimization for tree-based indices like Lucene. That deserves a test.

File: server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestUidMDCStorage.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.slf4j.MDC;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Wraps MDC calls to store the HTTP request UID in the {@link MDC} into an {@link AutoCloseable}.
+ */
+class RequestUidMDCStorage implements AutoCloseable {
+  private static final String HTTP_REQUEST_UID_MDC_KEY = "UID";
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestUidMDCStorage.java]: Why do you introduce the term "UID" while "UUID" and "ID" are commonly used ?

File: server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestUidFilter.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  public void init(FilterConfig filterConfig) throws ServletException {
+    // nothing to do
+  }
+
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+    RequestUidGenerator requestUidGenerator = platform.getContainer().getComponentByType(RequestUidGenerator.class);
+    RequestUidMDCStorageFactory requestUidMDCStorageFactory = platform.getContainer().getComponentByType(RequestUidMDCStorageFactory.class);
+
+    try (RequestUidMDCStorage mdcStorage = requestUidMDCStorageFactory.put(requestUidGenerator.generate())) {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestUidFilter.java]: Implementing a factory and a AutoCloseable resource is over-design. Multiplicity of classes does not help understanding how this "simple" (in term of amount of code) algorithm is implemented. Moreover no need for genericity here.

Following code is simpler to me:
```
public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
    RequestUidGenerator requestUidGenerator = platform.getContainer().getComponentByType(RequestUidGenerator.class);

    try {
      MDC.put(HTTP_REQUEST_UID_MDC_KEY, requestUidGenerator.generate());
      chain.doFilter(request, response);
    } finally {
      MDC.remove(HTTP_REQUEST_UID_MDC_KEY);
    }
  }
```

It drops 2 classes and a useless `requireNotNull` check in `RequestUidMDCStorage`.
KISS !

Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestUidFilter.java]: I went for a solution half way: I kept RequestUidMDCStorage to isolate code and responsibility but dropped the factory

File: server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestUidMDCStorage.java (hunk 9 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestUidMDCStorage.java]: UID dropped

**Comment by sns-seb** [comment]
> The ID is missing on logger named http :

this log will be removed in incoming ticket SONAR-8339

File: server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestIdGeneratorBase.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.platform.web.requestid;
+
+import org.sonar.core.util.UuidGenerator;
+
+public interface RequestIdGeneratorBase {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestIdGeneratorBase.java]: CRITICAL Annotate the "RequestIdGeneratorBase" interface with the @FunctionalInterface annotation [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1609)

File: server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestIdGenerator.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.platform.web.requestid;
+
+/**
+ * Generate a Unique Identifier for Http Requests.
+ */
+public interface RequestIdGenerator {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/platform/web/requestid/RequestIdGenerator.java]: CRITICAL Annotate the "RequestIdGenerator" interface with the @FunctionalInterface annotation [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1609)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* CRITICAL 2 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Pull request merged**